src/cirrus/lib2/events.py

```diff
@@ -0,0 +1,305 @@
+        self.event_publisher = (
+            SNSPublisher(
+                wf_event_topic_arn,
+                logger=self.logger,
+                boto3_session=self._boto3_session,
+                batch_size=batch_size,
+            )
+            if wf_event_topic_arn
+            else None
+        )
+        self.statedb = statedb if statedb is not None else StateDB.get_singleton()
+        self.eventdb = (
+            eventdb if eventdb is not None else EventDB(session=self._boto3_session)
+        )
+
+    def flush(self: "WorkflowEventManager"):
+        """Ensure any messages remaining in the batch buffer are sent."""
+        if self.event_publisher:
+            self.event_publisher.execute()
+
+    def __enter__(self: "WorkflowEventManager"):
+        return self
+
+    def __exit__(self: "WorkflowEventManager", et, ev, tb):
+        self.flush()
+
+    @classmethod
+    def with_wfem(cls: "WorkflowEventManager", *wfem_args, **wfem_kwargs):
+        """
+        Decorator to inject WorkflowEvenManager (as `wfem`) into a function call, and
+        flush upon exit.
+
+        Args:
+           See `help(WorkflowEvenManager.__init__)`
+
+        Returns:
+           A function which runs the wrapped function in a WorkflowEvenManager context,
+           and passes in the WorkflowEvenManager instance to be used within the the
+           function
+        """
+
+        def decorator(function: Callable):
+            @wraps(function)
+            def wrap_function(*args, **kwargs):
+                with cls(*wfem_args, **wfem_kwargs) as wfem:
+                    return function(*args, **kwargs, wfem=wfem)
+
+            return wrap_function
+
+        return decorator
+
+    def announce(
+        self: "WorkflowEventManager",
+        event_type: WFEventType,
+        payload_id: str,
+        payload_url: str = None,
+        extra_message: dict = None,
+    ) -> None:
+        """
+        Construct message payload and publish to WorkflowEventTopic.
+
+        Args:
+            event_type (WFEventType): the type of event which occurred
+            payload_id (str): The ID of the payload.
+            payload_url (str): URL of the ProcessPayload which triggered the event.
+            extra_message (dict): dictionary of additional items to be placed into the
+                                  SNS message body.
+        """
+        if self.event_publisher is None:
+            return
+        if payload_id is None:
+            raise ValueError("Must specify a payload_id")
+
+        message = {
+            "event_type": event_type,
+            "payload_url": payload_url,
+            "payload_id": payload_id,
+        }
+
+        if extra_message:
+            if not all(k not in message for k in extra_message):
+                raise ValueError(
+                    "extra_message parameters must not include: "
+                    + f"{','.join(message.keys())}."
+                )
```

Comment:
Thought for the future: I think it might be better to use something like a dataclass to more strongly type the event message format. `event_type`, `payload_url`, and `payload_id` would be required fields, and things like `error`, `execution_arn`, etc would be optional.

Then this `announce` method would only need to take a the event message type as an argument. We could also use that class as a place to centralize logic like serializing the message to json and creating the message attributes.